clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgent.groovy

```diff
@@ -58,10 +83,80 @@ class OpenstackSecurityGroupCachingAgent extends AbstractOpenstackCachingAgent {
      * that there are duplicate security groups by name, the lookup is to a list of ids.
      */
     List<SecGroupExtension> securityGroups = clientProvider.getSecurityGroups(region)
+    List<String> keys = securityGroups.collect{ Keys.getSecurityGroupKey(it.name, it.id, accountName, region) }
+
+    buildLoadDataCache(providerCache, keys) { CacheResultBuilder cacheResultBuilder ->
+      buildCacheResult(cacheResultBuilder, securityGroups)
+    }
+  }
+
+  @Override
+  OnDemandAgent.OnDemandResult handle(ProviderCache providerCache, Map<String, ? extends Object> data) {
+    log.debug("Handling on-demand cache update; account=${account}, region=${region}, data=${data}")
+
+    if (data.account != accountName) {
+      return null
+    }
+
+    if (data.region != region) {
+      return null
+    }
+
+    if (!data.containsKey('securityGroupName')) {
+      return null
+    }
+
+    String name = data.securityGroupName as String
+
+    SecGroupExtension securityGroup = metricsSupport.readData {
+      SecGroupExtension group = null
+      try {
+        /*
+         * Since we only have a name, we need to get all groups and filter by name. Also, since name is unique,
+         * ensure there is only security group by this name.
+         */
+        List<SecGroupExtension> groups = clientProvider.getSecurityGroups(region).findAll { it.name == name }
+        if (groups.size() == 1) {
+          group = groups.first()
+        } else {
+          log.warn("Failed to find unique security group with name ${name} in region ${region}")
+        }
+      } catch (OpenstackProviderException e) {
+        //Do nothing ... Exception is thrown if a security group isn't found
+        log.debug("Unable to find security group to add to OnDemand cache", e)
+      }
+      return group
+    }
+
+    List<SecGroupExtension> securityGroups = []
+    String key = Keys.getSecurityGroupKey(name, '*', accountName, region)
+    if (securityGroup) {
+      securityGroups = [securityGroup]
+      key = Keys.getSecurityGroupKey(name, securityGroup.id, accountName, region)
+    }
+
+    CacheResult cacheResult = metricsSupport.transformData {
+      buildCacheResult(new CacheResultBuilder(startTime: Long.MAX_VALUE), securityGroups)
+    }
+
+    String namespace = SECURITY_GROUPS.ns
+    String resolvedKey = null
+    try {
+      resolvedKey = resolveKey(providerCache, namespace, key)
+      processOnDemandCache(cacheResult, objectMapper, metricsSupport, providerCache, resolvedKey)
+    } catch(UnresolvableKeyException e) {
+      log.info("Security group ${name} is not resolvable", e)
```

Comment:
Wouldn't this cause the cache update to fail? Or really just never call `processOnDemandCache`?